src/feat/feature-mfcc.cc

```diff
@@ -44,22 +101,37 @@ Mfcc::Mfcc(const MfccOptions &opts)
   int32 padded_window_size = opts.frame_opts.PaddedWindowSize();
   if ((padded_window_size & (padded_window_size-1)) == 0)  // Is a power of two...
     srfft_ = new SplitRadixRealFft<BaseFloat>(padded_window_size);
-  
+
   // We'll definitely need the filterbanks info for VTLN warping factor 1.0.
-  // [note: this call caches it.]  The reason we call this here is to
-  // improve the efficiency of the "const" version of Compute().
+  // [note: this call caches it.]
   GetMelBanks(1.0);
 }
 
-Mfcc::~Mfcc() {
+MfccComputer::MfccComputer(const MfccComputer &other):
+    opts_(other.opts_), lifter_coeffs_(other.lifter_coeffs_),
+    dct_matrix_(other.dct_matrix_),
+    log_energy_floor_(other.log_energy_floor_),
+    mel_banks_(other.mel_banks_),
+    srfft_(NULL),
+    mel_energies_(other.mel_energies_.Dim(), kUndefined) {
+  for (std::map<BaseFloat, MelBanks*>::iterator iter = mel_banks_.begin();
+       iter != mel_banks_.end(); ++iter)
+    iter->second = new MelBanks(*(iter->second));
+  if (other.srfft_ != NULL)
+    srfft_ = new SplitRadixRealFft<BaseFloat>(*(other.srfft_));
+}
+
+
+
+MfccComputer::~MfccComputer() {
   for (std::map<BaseFloat, MelBanks*>::iterator iter = mel_banks_.begin();
       iter != mel_banks_.end();
       ++iter)
     delete iter->second;
   delete srfft_;
 }
 
-const MelBanks *Mfcc::GetMelBanks(BaseFloat vtln_warp) {
+const MelBanks *MfccComputer::GetMelBanks(BaseFloat vtln_warp) {
```

Comment:
The original code, that has been there for years, had a lot of repetition.
I don't really have time to remove all of it with this particular
refactoring- my goal was just to enable correct enforcement of the
snip-edges=false option.    (Anyway, in general I'm not very hawkish on
removing repetition).
Dan


On Mon, Apr 11, 2016 at 12:05 AM, Kirill Katsnelson <
notifications@github.com> wrote:

> In src/feat/feature-mfcc.cc
> <https://github.com/kaldi-asr/kaldi/pull/679#discussion_r59161400>:
>
> >    for (std::map<BaseFloat, MelBanks*>::iterator iter = mel_banks_.begin();
> >        iter != mel_banks_.end();
> >        ++iter)
> >      delete iter->second;
> >    delete srfft_;
> >  }
> >
> > -const MelBanks *Mfcc::GetMelBanks(BaseFloat vtln_warp) {
> > +const MelBanks *MfccComputer::GetMelBanks(BaseFloat vtln_warp) {
>
> A lot of copy-paste is going on here. Maybe a common base class can supply
> GetMelBanks() cache to multiple XxxComputer classes? Also, can provide
> computing power spectrum and an HTK compatible feature reordering utility
> methods that are also ubiquitous.
>
> With common options structure inheritance, can bring even more DRY
> <https://en.wikipedia.org/wiki/Don't_repeat_yourself> in registering
> command line options, as well as into the Compute() methods.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/679/files/8e7c759093ca984eeab029d0c6c1393249f1a158#r59161400>
>